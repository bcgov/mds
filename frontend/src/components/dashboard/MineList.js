import React, { Component } from "react";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";
import { Col, Row, Divider } from "antd";
import * as router from "@/constants/routes";
import NullScreen from "@/components/common/NullScreen";

/**
 * @class MineList - paginated list of mines
 */

const propTypes = {
  mines: PropTypes.object.isRequired,
  mineIds: PropTypes.array.isRequired,
  pageData: PropTypes.object.isRequired,
  mineRegionHash: PropTypes.object.isRequired,
};

const defaultProps = {
  mines: {},
  mineIds: [],
  pageData: {},
};

class MineList extends Component {
  render() {
    const { mines, mineIds, mineRegionHash } = this.props;
    return (
      <div>
        <Row type="flex" style={{ textAlign: "center" }}>
          <Col span={6}>
            <h2>Name</h2>
          </Col>
          <Col span={6}>
            <h2>Mine ID</h2>
          </Col>
          <Col span={6}>
            <h2>Region</h2>
          </Col>
          <Col span={6}>
            <h2>Permit(s)</h2>
          </Col>
        </Row>
        <Divider style={{ height: "2px", backgroundColor: "#013366", margin: "0" }} />
        {mineIds &&
          mineIds.map((id) => (
            <div key={id}>
              <Row type="flex" style={{ textAlign: "center" }}>
                <Col id="mine_list_name" span={6}>
                  <Link to={router.MINE_SUMMARY.dynamicRoute(id)}>
                    {mines[id].mine_detail[0]
                      ? mines[id].mine_detail[0].mine_name
                      : String.EMPTY_FIELD}
                  </Link>
                </Col>
                <Col id="mine_list_id" span={6}>
                  {mines[id].mine_detail[0] ? mines[id].mine_detail[0].mine_no : String.EMPTY_FIELD}
                </Col>
                <Col id="mine_list_region" span={6}>
                  {mines[id].mine_detail[0]
                    ? mineRegionHash[mines[id].mine_detail[0].region_code]
                    : String.EMPTY_FIELD}
                </Col>
                <Col id="mine_list_permit" span={6}>
                  {mines[id].mine_permit &&
                    mines[id].mine_permit.map((permit) => (
                      <div key={permit.permit_guid}>{permit.permit_no}</div>
                    ))}
                </Col>
                <Divider />
              </Row>
            </div>
          ))}
        {mineIds.length === 0 && <NullScreen type="no-results" />}
      </div>
    );
  }
}

MineList.propTypes = propTypes;
MineList.defaultProps = defaultProps;

export default MineList;
