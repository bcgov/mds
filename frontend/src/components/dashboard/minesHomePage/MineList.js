import React from "react";
import { objectOf, arrayOf, string } from "prop-types";
import { Link } from "react-router-dom";
import { Table } from "antd";
import { uniqBy } from "lodash";
import * as router from "@/constants/routes";
import * as Strings from "@/constants/strings";
import NullScreen from "@/components/common/NullScreen";
import CustomPropTypes from "@/customPropTypes";

/**
 * @class MineList - paginated list of mines
 */

const propTypes = {
  mines: objectOf(CustomPropTypes.mine).isRequired,
  mineIds: arrayOf(string).isRequired,
  mineRegionHash: objectOf(string).isRequired,
  mineTenureHash: objectOf(string).isRequired,
  mineCommodityOptionsHash: objectOf(string).isRequired,
};

const columns = [
  {
    title: "Mine Name",
    width: 200,
    dataIndex: "mineName",
    render: (text, record) => <Link to={router.MINE_SUMMARY.dynamicRoute(record.key)}>{text}</Link>,
  },
  {
    title: "Mine No.",
    width: 100,
    dataIndex: "mineNo",
    render: (text) => <div title="Mine Number">{text}</div>,
  },
  {
    title: "Operational Status",
    width: 150,
    dataIndex: "operationalStatus",
    render: (text) => <div title="Operational Status">{text}</div>,
  },
  {
    title: "Permit No.",
    dataIndex: "permit",
    width: 150,
    render: (text, record) => (
      <div title="Permit Number">
        <ul className="mine-list__permits">
          {text &&
            uniqBy(text, "permit_no").map(({ permit_no, permit_guid }) => (
              <li key={permit_guid}>{permit_no}</li>
            ))}
          {!text && <li>{record.emptyField}</li>}
        </ul>
      </div>
    ),
  },
  {
    title: "Region",
    dataIndex: "region",
    width: 150,
    render: (text, record) => (
      <div title="Region">
        {text}
        {!text && <div>{record.emptyField}</div>}
      </div>
    ),
  },
  {
    title: "Tenure",
    dataIndex: "tenure",
    width: 150,
    render: (text, record) => (
      <div title="Tenure">
        {text &&
          text.map((tenure) => (
            <span className="mine_tenure" key={tenure.mine_type_guid}>
              {record.tenureHash[tenure.mine_tenure_type_code]}
            </span>
          ))}
        {!text && <div>{record.emptyField}</div>}
      </div>
    ),
  },
  {
    title: "Commodity",
    dataIndex: "commodity",
    width: 150,
    render: (text, record) => (
      <div title="Commodity">
        {text &&
          text.map(({ mine_type_detail, mine_type_guid }) => (
            <div key={mine_type_guid}>
              {mine_type_detail.map(({ mine_commodity_code, mine_type_detail_guid }) => (
                <span key={mine_type_detail_guid}>
                  {mine_commodity_code && `${record.commodityHash[mine_commodity_code]},`}
                </span>
              ))}
            </div>
          ))}
      </div>
    ),
  },
  {
    title: "TSF",
    dataIndex: "tsf",
    width: 150,
    render: (text) => <div title="TSF">{text}</div>,
  },
];

const transformRowData = (mines, mineIds, mineRegionHash, mineTenureHash, mineCommodityHash) =>
  mineIds.map((id) => ({
    key: id,
    emptyField: Strings.EMPTY_FIELD,
    mineName: mines[id].mine_name ? mines[id].mine_name : Strings.EMPTY_FIELD,
    mineNo: mines[id].mine_no ? mines[id].mine_no : Strings.EMPTY_FIELD,
    operationalStatus: mines[id].mine_status[0]
      ? mines[id].mine_status[0].status_labels[0]
      : Strings.EMPTY_FIELD,
    permit: mines[id].mine_permit[0] ? mines[id].mine_permit : null,
    region: mines[id].region_code ? mineRegionHash[mines[id].region_code] : Strings.EMPTY_FIELD,
    commodity: mines[id].mine_type[0] ? mines[id].mine_type : null,
    commodityHash: mineCommodityHash,
    tenure: mines[id].mine_type[0] ? mines[id].mine_type : null,
    tenureHash: mineTenureHash,
    tsf: mines[id].mine_tailings_storage_facility
      ? mines[id].mine_tailings_storage_facility.length
      : Strings.EMPTY_FIELD,
  }));

export const MineList = (props) => (
  <Table
    align="left"
    pagination={false}
    columns={columns}
    dataSource={transformRowData(
      props.mines,
      props.mineIds,
      props.mineRegionHash,
      props.mineTenureHash,
      props.mineCommodityOptionsHash
    )}
    locale={{ emptyText: <NullScreen type="no-results" /> }}
  />
);

MineList.propTypes = propTypes;

export default MineList;
