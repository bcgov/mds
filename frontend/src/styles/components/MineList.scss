.mine-list table {
  min-width: 1500px;
}

.mine-list__permits {
  margin: 0;
  display: inline-block;
  vertical-align: top;
  padding: 0;
  list-style-type: none;
}