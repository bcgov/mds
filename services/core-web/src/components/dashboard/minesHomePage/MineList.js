import React from "react";
import { func, objectOf, arrayOf, string, bool } from "prop-types";
import { Link } from "react-router-dom";
import { isEmpty } from "lodash";
import * as Strings from "@common/constants/strings";
import * as router from "@/constants/routes";
import NullScreen from "@/components/common/NullScreen";
import CoreTable from "@/components/common/CoreTable";
import CustomPropTypes from "@/customPropTypes";
import { SUCCESS_CHECKMARK } from "@/constants/assets";

/**
 * @class MineList - paginated list of mines
 */

const propTypes = {
  mines: objectOf(CustomPropTypes.mine).isRequired,
  mineIds: arrayOf(string).isRequired,
  mineRegionHash: objectOf(string).isRequired,
  mineTenureHash: objectOf(string).isRequired,
  mineCommodityOptionsHash: objectOf(string).isRequired,
  handleMineSearch: func.isRequired,
  sortField: string,
  sortDir: string,
  isLoaded: bool.isRequired,
};

const defaultProps = {
  sortField: null,
  sortDir: null,
};

const columns = [
  {
    title: "Mine Name",
    dataIndex: "mineName",
    sortField: "mine_name",
    width: 150,
    render: (text, record) => (
      <Link to={router.MINE_SUMMARY.dynamicRoute(record.key)}>
        {text}
        {record.verified_status.healthy_ind && (
          <img
            alt="checkmark"
            className="padding-small"
            src={SUCCESS_CHECKMARK}
            width="25"
            title={`Mine data verified by ${record.verified_status.verifying_user}`}
          />
        )}
      </Link>
    ),
    sorter: true,
  },
  {
    title: "Mine No.",
    dataIndex: "mineNo",
    sortField: "mine_no",
    width: 150,
    render: (text) => <div title="Mine Number">{text}</div>,
    sorter: true,
  },
  {
    title: "Operational Status",
    dataIndex: "operationalStatus",
    sortField: "mine_operation_status_code",
    width: 150,
    render: (text) => <div title="Operational Status">{text}</div>,
  },
  {
    title: "Permit No.",
    dataIndex: "permit",
    width: 150,
    render: (text, record) => (
      <div title="Permit Number">
        <ul className="mine-list__permits">
          {text && text.map((permit_no) => <li key={permit_no}>{permit_no}</li>)}
          {!text && <li>{record.emptyField}</li>}
        </ul>
      </div>
    ),
  },
  {
    title: "Region",
    dataIndex: "region",
    sortField: "mine_region",
    width: 150,
    render: (text, record) => (
      <div title="Region">
        {text}
        {!text && <div>{record.emptyField}</div>}
      </div>
    ),
    sorter: true,
  },
  {
    title: "Tenure",
    dataIndex: "tenure",
    width: 150,
    render: (text, record) => (
      <div title="Tenure">
        {text &&
          text.map((tenure) => (
            <span className="mine_tenure" key={tenure.mine_type_guid}>
              {record.tenureHash[tenure.mine_tenure_type_code]}
            </span>
          ))}
        {!text && <div>{record.emptyField}</div>}
      </div>
    ),
  },
  {
    title: "Commodity",
    dataIndex: "commodity",
    width: 150,
    render: (text, record) => (
      <div title="Commodity">
        {(text &&
          text
            .map(({ mine_type_detail }) =>
              mine_type_detail
                .map(
                  ({ mine_commodity_code }) =>
                    mine_commodity_code && record.commodityHash[mine_commodity_code]
                )
                .filter(Boolean)
                .join(", ")
            )
            .join(", ")) ||
          record.emptyField}
      </div>
    ),
  },
  {
    title: "TSF",
    dataIndex: "tsf",
    width: 150,
    render: (text) => <div title="TSF">{text}</div>,
  },
];

const transformRowData = (mines, mineIds, mineRegionHash, mineTenureHash, mineCommodityHash) =>
  mineIds
    .filter((id) => mines[id])
    .map((id) => ({
      key: id,
      emptyField: Strings.EMPTY_FIELD,
      mineName: mines[id].mine_name ? mines[id].mine_name : Strings.EMPTY_FIELD,
      mineNo: mines[id].mine_no ? mines[id].mine_no : Strings.EMPTY_FIELD,
      operationalStatus:
        mines[id].mine_status &&
        mines[id].mine_status[0] &&
        mines[id].mine_status[0].status_labels &&
        mines[id].mine_status[0].status_labels[0]
          ? mines[id].mine_status[0].status_labels[0]
          : Strings.EMPTY_FIELD,
      permit:
        mines[id].mine_permit_numbers && mines[id].mine_permit_numbers[0]
          ? mines[id].mine_permit_numbers
          : null,
      region: mines[id].mine_region ? mineRegionHash[mines[id].mine_region] : Strings.EMPTY_FIELD,
      commodity: mines[id].mine_type && mines[id].mine_type[0] ? mines[id].mine_type : null,
      commodityHash: mineCommodityHash,
      tenure: mines[id].mine_type && mines[id].mine_type[0] ? mines[id].mine_type : null,
      tenureHash: mineTenureHash,
      tsf: mines[id].mine_tailings_storage_facilities
        ? mines[id].mine_tailings_storage_facilities.length
        : Strings.EMPTY_FIELD,
      verified_status: mines[id].verified_status,
    }));

const handleTableChange = (updateMineList) => (pagination, filters, sorter) => {
  const params = isEmpty(sorter)
    ? {
        sort_field: undefined,
        sort_dir: undefined,
      }
    : {
        sort_field: sorter.column.sortField,
        sort_dir: sorter.order.replace("end", ""),
      };
  updateMineList(params);
};

const applySortIndicator = (_columns, field, dir) =>
  _columns.map((column) => ({
    ...column,
    sortOrder: column.sortField === field ? dir.concat("end") : false,
  }));

export const MineList = (props) => (
  <CoreTable
    condition={props.isLoaded}
    columns={applySortIndicator(columns, props.sortField, props.sortDir)}
    dataSource={transformRowData(
      props.mines,
      props.mineIds,
      props.mineRegionHash,
      props.mineTenureHash,
      props.mineCommodityOptionsHash
    )}
    tableProps={{
      align: "left",
      pagination: false,
      locale: { emptyText: <NullScreen type="no-results" /> },
      onChange: handleTableChange(props.handleMineSearch),
    }}
  />
);

MineList.propTypes = propTypes;
MineList.defaultProps = defaultProps;

export default MineList;
