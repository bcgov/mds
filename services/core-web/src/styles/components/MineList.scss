.mine-list__permits {
  margin: 0;
  display: inline-block;
  vertical-align: top;
  padding: 0;
  list-style-type: none;
}
