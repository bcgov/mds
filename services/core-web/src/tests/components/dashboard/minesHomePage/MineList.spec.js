import React from "react";
import { shallow } from "enzyme";
import MineList from "@/components/dashboard/minesHomePage/MineList";
import * as MOCK from "@/tests/mocks/dataMocks";

const props = {};

const setupProps = () => {
  props.mines = MOCK.MINES.mines;
  props.mineRegionHash = MOCK.REGION_HASH;
  props.mineTenureHash = MOCK.TENURE_HASH;
};

beforeEach(() => {
  setupProps();
});

describe("MineList", () => {
  it("renders properly", () => {
    const component = shallow(<MineList {...props} />);
    expect(component).toMatchSnapshot();
  });
});
