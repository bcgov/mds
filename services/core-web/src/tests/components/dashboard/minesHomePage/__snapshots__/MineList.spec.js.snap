// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MineList renders properly 1`] = `
<CoreTable
  columns={
    Array [
      Object {
        "dataIndex": "mineName",
        "render": [Function],
        "sortField": "mine_name",
        "sortOrder": false,
        "sorter": true,
        "title": "Mine Name",
        "width": 150,
      },
      Object {
        "dataIndex": "mineNo",
        "render": [Function],
        "sortField": "mine_no",
        "sortOrder": false,
        "sorter": true,
        "title": "Mine No.",
        "width": 150,
      },
      Object {
        "dataIndex": "operationalStatus",
        "render": [Function],
        "sortField": "mine_operation_status_code",
        "sortOrder": false,
        "title": "Operational Status",
        "width": 150,
      },
      Object {
        "dataIndex": "permit",
        "render": [Function],
        "sortOrder": false,
        "title": "Permit No.",
        "width": 150,
      },
      Object {
        "dataIndex": "region",
        "render": [Function],
        "sortField": "mine_region",
        "sortOrder": false,
        "sorter": true,
        "title": "Region",
        "width": 150,
      },
      Object {
        "dataIndex": "tenure",
        "render": [Function],
        "sortOrder": false,
        "title": "Tenure",
        "width": 150,
      },
      Object {
        "dataIndex": "commodity",
        "render": [Function],
        "sortOrder": false,
        "title": "Commodity",
        "width": 150,
      },
      Object {
        "dataIndex": "tsf",
        "render": [Function],
        "sortOrder": false,
        "title": "TSF",
        "width": 150,
      },
    ]
  }
  dataSource={
    Array [
      Object {
        "commodity": Array [
          Object {
            "mine_tenure_type_code": "PLR",
          },
          Object {
            "mine_tenure_type_code": "MIN",
          },
        ],
        "commodityHash": undefined,
        "emptyField": "N/A",
        "key": "18133c75-49ad-4101-85f3-a43e35ae989a",
        "mineName": "mine3",
        "mineNo": "BLAH9091",
        "operationalStatus": "N/A",
        "permit": null,
        "region": "North East",
        "tenure": Array [
          Object {
            "mine_tenure_type_code": "PLR",
          },
          Object {
            "mine_tenure_type_code": "MIN",
          },
        ],
        "tenureHash": undefined,
        "tsf": 1,
        "verified_status": Object {
          "healthy_ind": null,
          "mine_guid": null,
          "mine_name": null,
          "verifying_timestamp": null,
          "verifying_user": null,
        },
      },
      Object {
        "commodity": Array [
          Object {
            "mine_tenure_type_code": "PLR",
          },
          Object {
            "mine_tenure_type_code": "MIN",
          },
        ],
        "commodityHash": undefined,
        "emptyField": "N/A",
        "key": "18145c75-49ad-0101-85f3-a43e45ae989a",
        "mineName": "mine2",
        "mineNo": "BLAH9091",
        "operationalStatus": "N/A",
        "permit": Array [
          "KNaaWwVdiJ40",
          "xfM0c0ZKEw7B",
        ],
        "region": "North East",
        "tenure": Array [
          Object {
            "mine_tenure_type_code": "PLR",
          },
          Object {
            "mine_tenure_type_code": "MIN",
          },
        ],
        "tenureHash": undefined,
        "tsf": 1,
        "verified_status": Object {
          "healthy_ind": true,
          "mine_guid": "18145c75-49ad-0101-85f3-a43e45ae989a",
          "mine_name": "mine2",
          "verifying_timestamp": null,
          "verifying_user": null,
        },
      },
    ]
  }
  tableProps={
    Object {
      "align": "left",
      "locale": Object {
        "emptyText": <NullScreen
          message=""
          type="no-results"
        />,
      },
      "onChange": [Function],
      "pagination": false,
    }
  }
/>
`;
